app/src/main/java/com/fivegmag/a5gmsmediasessionhandler/MediaSessionHandlerMessengerService.kt

```diff
@@ -302,5 +323,177 @@ class MediaSessionHandlerMessengerService() : Service() {
+        {
+            Log.i(TAG, "[ConsumptionReporting] IsConsumptionReportingActivated: report triggered by play status【${state}】 v2")
+            clientsSessionData[clientId]!!.playbackState = PlayerStates.UNKNOWN
+            return true
+        }
+
+        // Condition 3: check clientConsumptionReportingConfiguration.reportingInterval, timer trigger
+        if (clientsSessionData[clientId]!!.isConsumptionReportByTimer)
+        {
+            Log.i(TAG, "[ConsumptionReporting] IsConsumptionReportingActivated: report triggered by timer v2")
+
+            clientsSessionData[clientId]!!.isConsumptionReportByTimer = false
+            return true
+        }
+
+        // Condition 4&5:check clientConsumptionReportingConfiguration.locationReporting and clientConsumptionReportingConfiguration.accessReporting
+        if(clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.locationReporting
+            || clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.accessReporting)
+        {
+            Log.i(TAG, "[ConsumptionReporting] IsConsumptionReportingActivated: report triggered by locationReporting/accessReporting")
+            return true
+        }
+
+        return false
+    }
+
+    private fun reportConsumption(msg: Message) {
+        val sendingUid = msg.sendingUid;
+        if (!isNeedReportConsumption(sendingUid))
+        {
+            Log.i(TAG, "[ConsumptionReporting] Not need ReportConsumption")
+            return
+        }
+
+        val bundle: Bundle = msg.data
+        bundle.classLoader = ConsumptionReporting::class.java.classLoader
+        val dataReporting: ConsumptionReporting? = bundle.getParcelable("consumptionData")
+
+        Log.i(TAG, "[ConsumptionReporting] reportConsumption: ClientId[${dataReporting?.reportingClientId}] , " +
+                "Entry[${dataReporting?.mediaPlayerEntry}], " +
+                "startTime[${dataReporting?.consumptionReportingUnits?.get(0)?.startTime}]," +
+                "duration[${dataReporting?.consumptionReportingUnits?.get(0)?.duration}]," +
+                "ipAddr[${dataReporting?.consumptionReportingUnits?.get(0)?.mediaEndpointAddress?.ipv4Addr}/" +
+                "${dataReporting?.consumptionReportingUnits?.get(0)?.mediaEndpointAddress?.ipv6Addr}]"
+        )
+
+        Toast.makeText(
+            applicationContext,
+            "MSH recv Consumption-ID: ${dataReporting?.reportingClientId}",
+            Toast.LENGTH_LONG
+        ).show()
+
+        // call m5 report consumption to AF - TS26.512 Clause 4.7.4
+        val consumptionReportingApi: ConsumptionReportingApi? = clientsSessionData[sendingUid]?.consumptionReportingApi
+        if(consumptionReportingApi == null)
+        {
+            Log.i(TAG, "[ConsumptionReporting] consumptionReportingApi is 【NULL】")
+            return
+        }
+
+        val provisisioningSessionId: String = "2";
```

Comment:
Why is this value hardcoded?

Reply:
I didn't found the defintion of provisisioningSessionId in TS26.512, I didn't know how to generate it, so I hardcoded first.

Reply:
Note the parameter is called `provisioningSessionId`, you have typos throughout your code.

Reply:
Got it. Thanks